ignite/metrics/metric.py

```diff
@@ -0,0 +1,66 @@
+from abc import ABCMeta, abstractmethod
+
+from ignite.engine import Events
+
+
+class Metric(object):
+    __metaclass__ = ABCMeta
+
+    """
+    Base class for all Metrics.
+
+    Metrics provide a way to compute various quantities of interest in an online
+    fashion without having to store the entire output history of a model.
+    """
+    def __init__(self):
+        self.reset()
+
+    @abstractmethod
+    def reset(self):
+        """
+        Resets the metric to to it's initial state.
+
+        This is called at the start of each evaluation run.
+        """
+        pass
+
+    @abstractmethod
+    def update(self, output):
+        """
+        Updates the metric's state using the passed batch output.
+
+        This is called once for each batch of each evaluation run.
+
+        Args:
+            output: the is the output from the evaluator's process function
+        """
+        pass
+
+    @abstractmethod
+    def compute(self):
+        """
+        Computes the metric based on it's accumulated state.
+
+        This is called at the end of each evaluation run.
+
+        Returns:
+            Any: the actual quantity of interest
+
+        Raises:
+            NotComputableError: raised when the metric cannot be computed
+        """
+        pass
+
+    def started(self, engine, state):
+        self.reset()
+
+    def iteration_completed(self, engine, state):
+        self.update(state.output)
+
+    def completed(self, engine, state):
+        state.metrics.append(self.compute())
```

Comment:
not sure this is the best approach? What if we have several metrics, `state.metrics` should probably be a dict which is accessed with the name of the metric? which can be passed in to the constructor (and have a default value)?

Reply:
Yea I wasn't sure on this one. I'm def open to the dict approach. The only reason I went with the list is for the simpler API, both in the constructor and for accessing the result. I see two options if we want to use a more explicit approach:

Option 1
```python
evaluator = create_supervised_evaluator(model, metrics={'acc': CategoricalAccuracy()})
```

Option 2
```python
evaluator = create_supervised_evaluator(model, metrics=[CategoricalAccuracy(name='acc')])
```

What do you think?

Reply:
maybe option 1? then it's consistent with how you pass it in and how its represented on the `State`?

Reply:
👍 I think that sounds great. I'll make the change.